From b2868cdd2b8e8d4ebb059ee6c6586d30fd40e68b Mon Sep 17 00:00:00 2001
From: Marek Vasut <marex@nabladev.com>
Date: Tue, 7 Oct 2025 12:48:36 +0200
Subject: [PATCH 91/95] drm/bridge: synopsys: dw-mipi-dsi: Force LP11 on attach

Force the DSI data lanes into LP11 mode on attach by starting
the bridge. This allows bridges like TC358767 to start properly
without dropping into debug modes.

Upstream-Status: Pending
Signed-off-by: Marek Vasut <marex@nabladev.com>
---
 drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c b/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
index 6d5bcdc4e8e8f..cf6e1e92d47ca 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-mipi-dsi.c
@@ -1142,13 +1142,27 @@ static int dw_mipi_dsi_bridge_attach(struct drm_bridge *bridge,
 				     enum drm_bridge_attach_flags flags)
 {
 	struct dw_mipi_dsi *dsi = bridge_to_dsi(bridge);
+	const struct dw_mipi_dsi_phy_ops *phy_ops = dsi->plat_data->phy_ops;
+	int ret;
 
 	/* Set the encoder type as caller does not know it */
 	encoder->encoder_type = DRM_MODE_ENCODER_DSI;
 
+	/* Enable the bridge so the data lanes are in LP11 */
+	dw_mipi_dsi_mode_set(dsi, &dsi->mode);
+
 	/* Attach the panel-bridge to the dsi bridge */
-	return drm_bridge_attach(encoder, dsi->panel_bridge, bridge,
-				 flags);
+	ret = drm_bridge_attach(encoder, dsi->panel_bridge, bridge, flags);
+
+	/* Disable the bridge */
+	dw_mipi_dsi_set_mode(dsi, 0);
+	if (phy_ops->power_off)
+		phy_ops->power_off(dsi->plat_data->priv_data);
+	dw_mipi_dsi_disable(dsi);
+	clk_disable_unprepare(dsi->pclk);
+	pm_runtime_put(dsi->dev);
+
+	return ret;
 }
 
 static const struct drm_bridge_funcs dw_mipi_dsi_bridge_funcs = {
-- 
2.51.0

